flow/directives.py

```diff
@@ -227,17 +178,14 @@ def evaluate(self, jobs):
                 self[key] = _evaluate(value, jobs)
             self._evaluated = True
 
-    def _aggregate(self, other, jobs=None, parallel=False):
-        self.evaluate(jobs)
-        other.evaluate(jobs)
-        agg_func_attr = "_parallel" if parallel else "_serial"
-        for name in self._defined_directives:
-            agg_func = getattr(self._directive_definitions[name], agg_func_attr)
-            default_value = self._directive_definitions[name]._default
-            other_directive = other.get(name, default_value)
-            directive = self[name]
-            if other_directive is not None:
-                self._defined_directives[name] = agg_func(directive, other_directive)
+        directives = dict(self)
+        directives["cpus"] = directives["processes"] * directives["threads_per_process"]
+        directives["gpus"] = directives["processes"] * directives["gpus_per_process"]
+        if (memory := directives["memory_per_cpu"]) is not None:
+            directives["memory"] = directives["cpus"] * memory
+        else:
```

Comment:
`memory_per_gpu` is not a directive. In this new scheme `processes` is the base unit on which `gpus` and `memory` are based. If your users need a specific amount of memory per GPU, they could set:

```
directives={'processes': N, 'gpus_per_process': G, 'memory_per_process': memory_per_gpu * G}
```

Reply:
Wait, I got that wrong. The directive is (currently) `memory_per_cpu`, so you would have to calculate that based on the number of cpus per gpu.

I originally suggested `memory_per_cpu` because that provided a 1:1 mapping to SLURM. However, @b-butler explained that the complexities of bundling and groups essentially requires computing an aggregate `memory` and the dividing to issue the proper SLURM request. Thus, we could change the user facing directive to `memory_per_process`. 

In either case, the information will be there for the Georgia Tech's template to produce `--memory-per-gpu` if that is a hard requirement.

Reply:
`--memory-per-gpu`  is a variable in the Georgia Tech's submission script.  We can enter it in the custom any custom value as log as we can import the in the header block values (example https://github.com/bcrawford39GT/signac_numpy_tutorial/blob/main/signac_numpy_tutorial/project/templates/phoenix.sh), without Signac auto placing a default `--mem` option that can not be removed,   This is the main issue, meaning if we add anything the default `--mem` option is also added, causing 2 conflicting variables.  Maybe there can be a variable to not autoprint `--mem`?

I think the `memory_per_process` option would be fine instead of `--memory-per-cpu` and `--memory-per-gpu`, as long as they can import the value in the header block values (example https://github.com/bcrawford39GT/signac_numpy_tutorial/blob/main/signac_numpy_tutorial/project/templates/phoenix.sh) for manipulation if needed, and Signac does not write `--mem` automatically.  Alternatively, maybe you are saying `memory_per_process` will be automatically written to `--mem` ?

If original `--memory-per-cpu` and `--memory-per-gpu` it is not feasible to do that is fine.  I was under the impression it was OK, not an issue, and in progress.

Reply:
The new default template will always write `--mem-per-cpu` to the slurm file, not `--mem`.  We can put this in a `{% block memory %}` so that child scripts can override it if needed.

Reply:
OK.  As long as the user has an option to stop the default writing of `--mem-per-cpu` also that should be fine.  We just can not specify it 2x+ or it will error out as overspecified

Reply:
I wish we didn't need memory at all, but we do. A number of clusters that flow supports default to a ridiculously small amount of memory per rank. Users therefore need the ability to request more. It is the responsibility of the user to request an appropriate amount that does not result in extra charges.

The systems that flow needs to both a) combine memory requests across bundles/groups and b) provide the --mem-per-cpu needed for srun to correctly launch jobs are complex.

Reply:
I agree.  We can work with the final solution that works for everyone.  Just wanted to state the barriers to things, so you are aware.

Reply:
I wish Memory was free also, on GPUs too!  No need to buy commercial cards then